### PR TITLE
Cross compile Scala code using Gradle variants

### PR DESCRIPTION
**Pull request by jbaiera**
This PR introduces Gradle variants into the ES-Hadoop build in order to compile and test Scala code across different Scala versions instead of recursively invoking the build from within itself. This PR should allow us to upgrade to Scala 2.12 and above as well as newer versions of Spark without breaking release artifact contracts outside of a major release.

Also included is the removal of the shared core source directories for the Spark integration. This shared source has been refactored into a core project that is cross compiled for the supported Spark and Scala versions used in the SQL integration.

More detail on why this change has been needed can be found at https://github.com/elastic/elasticsearch-hadoop/issues/1423 as well as much of the work that has been leading up to this point.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/hadoop/gradle/fixture/ElasticsearchFixturePlugin.groovy [hunk 1]
```diff
@@ -39,16 +40,18 @@ class ElasticsearchFixturePlugin implements Plugin<Project> {
...
         def clustersContainer = project.extensions.getByName(TestClustersPlugin.EXTENSION_NAME) as NamedDomainObjectContainer<ElasticsearchCluster>
         def integTestCluster = clustersContainer.create("integTest") { ElasticsearchCluster cluster ->
             cluster.version = version
             cluster.testDistribution = TestDistribution.DEFAULT
         }
 
-        integrationTest.useCluster(integTestCluster)
-        // Add the cluster HTTP URI as a system property which isn't tracked as a task input
-        integrationTest.jvmArgumentProviders.add({ ["-Dtests.rest.cluster=${integTestCluster.httpSocketURI}"] } as CommandLineArgumentProvider)
+        integrationTests.all { StandaloneRestIntegTestTask integrationTest ->
```
Comment (breskeby) [inline buildSrc/src/main/groovy/org/elasticsearch/hadoop/gradle/fixture/ElasticsearchFixturePlugin.groovy:50]: side note: You might want to consider moving your builds to use the task avoidance api for some better performance. see https://docs.gradle.org/current/userguide/task_configuration_avoidance.html for details

File: buildSrc/src/main/groovy/org/elasticsearch/hadoop/gradle/fixture/ElasticsearchFixturePlugin.groovy [hunk 2]
```diff
@@ -29,7 +30,7 @@ class ElasticsearchFixturePlugin implements Plugin<Project> {
         // Optionally allow user to disable the fixture
         def useFixture = Boolean.parseBoolean(project.findProperty("tests.fixture.es.enable") ?: "true")
 
-        def integrationTestTask = project.tasks.getByName("integrationTest") as StandaloneRestIntegTestTask
+        def integrationTestTask = project.tasks.withType(StandaloneRestIntegTestTask)
```
Comment (breskeby) [inline buildSrc/src/main/groovy/org/elasticsearch/hadoop/gradle/fixture/ElasticsearchFixturePlugin.groovy:33]: the name is a bit misleading. it should be `integrationTestTasks` shouldn't it?

File: buildSrc/src/main/groovy/org/elasticsearch/hadoop/gradle/BuildPlugin.groovy [hunk 3]
```diff
@@ -365,12 +521,25 @@ class BuildPlugin implements Plugin<Project>  {
         // Co-locate all build artifacts into distributions subdir for easier build automation
         distribution.doLast {
             project.copy { CopySpec spec ->
```
Comment (breskeby) [inline buildSrc/src/main/groovy/org/elasticsearch/hadoop/gradle/BuildPlugin.groovy:523]: You should declare the inputs  / outputs of that task to allow gradle to detect it's up-to-date if nothing has changed. Probably better use a simple task of type `Copy` here

File: spark/core/build.gradle [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    all { SparkVariantPlugin.SparkVariant variant ->
+
+        String scalaCompileTaskName = project.sourceSets
+                .getByName(variant.getSourceSetName("main"))
+                .getCompileTaskName("scala")
+
+        project.configurations {
+            create(variant.configuration('embedded')) {
+                transitive = false
+                canBeResolved = true
```
Comment (breskeby) [inline spark/core/build.gradle:26]: do you need this to be resolved directly? aren't other configurations not just extending this one?

**Review by breskeby — APPROVED** [review]
lgtm. just added a few minor remarks though

**Comment by adriancole** [comment]
hoping this isn't stalled out. the only thing pinning Zipkin's spark to old scala and old spark is this. Is there any testing or other things outsiders can help with?

**Comment by Tagar** [comment]
@jbaiera @breskeby @mark-vieira with all the new commits does ES-Hadoop shaping up to be complete to cross-compile across different Scala/Gradle variants. Thx!!

**Comment by jbaiera** [comment]
@Tagar I am indeed picking this back up again, yes. This specifically should be good to merge now. I'll see about getting the Scala 2.12 support together soon, and hopefully get the efforts for Spark 3.0 revived/updated.

**Comment by Tagar** [comment]
thanks for the update @jbaiera ! that's great news

**Pull request merged**